[Enhancement] Use basic table in show statistics

eshishki opened the pull request:
## Why I'm doing:

`SHOW ANALYZE JOB` and `SHOW ANALYZE STATUS` are extremely slow on environments with external catalog analyze jobs.

`ShowAnalyzeJobStmt` and `ShowAnalyzeStatusStmt` call `metadataMgr.getTable()` once per output row for the privilege check and to render the table name. For analyze entries on external catalogs (Iceberg, Hive, Hudi, JDBC, ...) `getTable()` triggers a full metadata fetch per row — one network round trip for every line in the result.

This is the same N+1 problem that #70197 fixed for `information_schema.tables_config` via `metadataMgr.getBasicTable()`. The downstream `Authorizer.checkAnyActionOnTableLikeObject` already accepts the `BasicTable` interface directly — no cast needed.

## What I'm doing:

Switch both call sites to `metadataMgr.getBasicTable(...)`, which is the lightweight variant designed to avoid network interaction with external services (see `BasicTable` javadoc).

Honour the existing config flag `enable_external_catalog_information_schema_tables_access_full_metadata` introduced in #70197 so operators keep a single switch for opting in to full external metadata fetches across all `information_schema` / `SHOW` paths.

For the internal catalog `getBasicTable(...)` still returns a full `Table`, so the legacy `Columns` rendering (collapse to `"ALL"` when the explicit column list equals the full schema) is preserved via an `instanceof Table` cast. For external catalogs no schema is available without network IO, so the captured column names are always rendered.

## Behavior change (external catalogs only)

This PR intentionally changes two visible behaviors for external-catalog rows in `SHOW ANALYZE JOB` / `SHOW ANALYZE STATUS`:

1. Rows for tables that have been **deleted in the source catalog** (Hive/Iceberg) are no longer silently filtered out. The old code relied on `getTable()` returning `null` for missing external tables, throwing `SemanticException`, and the catch returning `null`. The new lightweight path cannot detect deletion without doing the very network IO this PR avoids, so such rows are now rendered with whatever name/columns were captured at analyze time.
2. When the user explicitly listed **all** columns in `ANALYZE TABLE ... (c1,...,cn)` on an external catalog, the `Columns` field used to collapse to `"ALL"`; it now always renders the captured names. Same root cause — we don't have the schema on the lightweight path.

Internal-catalog behavior is unchanged.

Fixes #47295
Related: #70197

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/sql/ast/ShowAnalyzeJobStmt.java:
```diff
@@ -80,13 +80,13 @@ public static List<String> showAnalyzeJobs(ConnectContext context,
                 String tableName = analyzeJob.getTableName();
-                Table table = GlobalStateMgr.getCurrentState().getMetadataMgr().getTable(analyzeJob.getCatalogName(),
-                        dbName, tableName);
+                BasicTable table = GlobalStateMgr.getCurrentState().getMetadataMgr().getBasicTable(
```
This interface always returns a table, whether it exists or not

eshishki replied:
we are doing it just to take table name and to check permissions
the table existed when we created a job

anyway happy to amend it if you say how

eshishki committed:
[Enhancement] Use BasicTable in SHOW ANALYZE JOB/STATUS to avoid network IO

ShowAnalyzeJobStmt and ShowAnalyzeStatusStmt call metadataMgr.getTable() per
row when rendering output. For analyze entries on external catalogs (Iceberg,
Hive, Hudi, JDBC, ...) this triggers a full metadata fetch per row — one
network round trip for every line in SHOW ANALYZE JOB / SHOW ANALYZE STATUS,
which makes these commands very slow on environments with active external
analyze jobs.

The downstream Authorizer.checkAnyActionOnTableLikeObject and the name/columns
output only need the BasicTable interface, which is the lightweight variant
designed to avoid network interaction with external services (see BasicTable
javadoc).

Switch both call sites to metadataMgr.getBasicTable(...) and honour the
existing config flag enable_external_catalog_information_schema_tables_access_full_metadata
introduced in #70197 so operators retain a single switch for opting in to
full external metadata fetches.

Fixes #47295
Related: #70197

Signed-off-by: Evgeniy Shishkin <eshishki@gmail.com>

eshishki commented:
@kevincai PTAL

eshishki commented:
@alvin-celerdata could you take a look?

This is a small follow-up to #70197 — same `getBasicTable` + `enable_external_catalog_information_schema_tables_access_full_metadata` flag pattern, applied to `SHOW ANALYZE JOB/STATUS` (which `#70197` did not touch). On environments with active external-catalog analyze jobs, these `SHOW` commands currently do one full external metadata fetch per output row.

CelerData-Reviewer commented:
@codex review

eshishki commented on fe/fe-core/src/main/java/com/starrocks/sql/ast/ShowAnalyzeStatusStmt.java:
```diff
@@ -68,8 +69,7 @@ public static List<String> showAnalyzeStatus(ConnectContext context,
 
-        long totalCollectColumnsSize = StatisticUtils.getCollectibleColumns(table).size();
-        if (null != columns && !columns.isEmpty() && (columns.size() != totalCollectColumnsSize)) {
+        if (null != columns && !columns.isEmpty()) {
```
need reviewer guidance on this one as basicTable does not return column list or column number and this change is somewhat behaviour change

eshishki commented:
@stephen-shelby — heads up on a subtle behavior change this PR introduces, surfaced by the failing `AnalyzeStmtTest.testShow`.

**Before:** `showAnalyzeStatus` called `metadataMgr.getTable(...)`. For an `AnalyzeStatus` whose external table had been deleted in the source catalog (Hive/Iceberg), `getTable` returned `null`, the code threw `SemanticException`, caught it, and returned `null` — so stale rows for deleted external tables were silently skipped in `SHOW ANALYZE STATUS`.

**After this PR:** `getBasicTable(..., fetchExternalMetadata=false)` returns a lightweight `ExternalCatalogTableBasicInfo` stub **without** touching the connector. The stub is non-null as long as the catalog itself exists, so the existence check no longer fires and `SHOW ANALYZE STATUS` will now render a row for a deleted external table (using whatever name/columns were captured at analyze time).

The failing test pins exactly this fail-soft behavior:
```java
AnalyzeStatus extenalAnalyzeStatus = new ExternalAnalyzeStatus(-1, "hive0", "partitioned_db",
        "tx", "tx:xxxx", ...);
Assertions.assertNull(ShowAnalyzeStatusStmt.showAnalyzeStatus(getConnectContext(), extenalAnalyzeStatus));
```
(The test's `NullPointerException` on `getStatus()` is downstream — in production `setStatus(...)` is always called right after construction in `StmtExecutor` / `*AnalyzeJob` / `ConnectorAnalyzeTask` before the status is ever rendered, so this NPE is test-only.)

The whole point of this PR is to avoid per-row network IO **for external catalogs**, so falling back to `getTable(...)` for external would defeat the entire optimization. Realistically the only sane option is to accept the behavior change: update the test to set a status and assert the rendered row, flip the PR template to "Yes, change in behavior" → Miscellaneous, and document that deleted external tables are no longer silently filtered from `SHOW ANALYZE STATUS`.

OK to go that route, or do you see a third way?

eshishki commented:
Update — pushed a follow-up amend that:

1. **Restores the legacy `Columns` "ALL" rendering for the internal catalog** via an `instanceof Table` cast. `getBasicTable(...)` returns a real `Table` for internal catalogs, so `StatisticUtils.getCollectibleColumns(...).size()` / `getBaseSchema().size()` is still available there. External catalogs only get a `BasicTable` stub (no schema, no network IO), so the captured column names are always rendered for them.
2. **Updates the PR body** to flip "Yes, change in behavior" + Miscellaneous, and documents the external-only behavior change (deleted external tables are no longer silently filtered; explicit full column list always renders the names).

The failing `AnalyzeStmtTest.testShow` will still need its `extenalAnalyzeStatus` sub-case updated (set a status, assert the rendered row instead of `null`) once the direction is acknowledged. Happy to push the test update in this PR if you confirm the behavior change is acceptable.

HangyuanLiu commented:
thx, you should add ut to coverage this enhancement

eshishki commented:
> thx, you should add ut to coverage this enhancement

@HangyuanLiu This PR changes behaviour for external table case and i need ack from maintainers to change existing UT which is failing now
